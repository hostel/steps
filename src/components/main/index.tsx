import React, { useState, useCallback } from "react";

import { Indicators } from "@components/indicators";

import styles from "./main.module.scss";

const data = [
  {
    title: "Design",
  },
  {
    title: "Build",
  },
  {
    title: "Launch",
  },
];

export const Main = (): React.ReactElement => {
  const [activeStep, setActiveStep] = useState(0);

  /**
   * Handler on click on steps for change active step
   */
  const onClick = useCallback(
    ({
      currentTarget: { dataset },
    }: React.MouseEvent<HTMLButtonElement, MouseEvent>) => {
      const indexAsNumber = Number(dataset.index);

      const isPossibleNext = indexAsNumber === activeStep + 1;
      const isPossiblePrev = indexAsNumber === activeStep - 1;

      if (dataset.index && (isPossibleNext || isPossiblePrev)) {
        setActiveStep(indexAsNumber);
      }
    },
    [activeStep]
  );

  return (
    <div className={styles.wrap}>
      <div className={styles.default}>
        <Indicators onClick={onClick} steps={data} activeStep={activeStep} />
      </div>
    </div>
  );
};
