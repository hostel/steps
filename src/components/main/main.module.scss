.wrap {
  margin: 100px;
}

.default {
  max-width: 740px;
}
