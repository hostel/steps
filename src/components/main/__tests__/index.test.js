import { render, screen, fireEvent } from "@testing-library/react";

import { Main } from "@components/main";

test("should render default active step (first step)", () => {
  render(<Main />);

  const activeStep = screen.getByRole("button", { pressed: true });

  expect(activeStep).toMatchSnapshot();
});

test("should render first step (b/c clicked last element)", () => {
  render(<Main />);

  const steps = screen.getAllByRole("button", { pressed: false });

  const lastStep = steps[steps.length - 1];

  fireEvent.click(lastStep);

  const activeSteps = screen.getAllByRole("button", { pressed: true });

  expect(activeSteps).toMatchSnapshot();
});

test("should render first and second steps", () => {
  render(<Main />);

  const steps = screen.getAllByRole("button", { pressed: false });

  const secondStep = steps[0];

  fireEvent.click(secondStep);

  const activeSteps = screen.getAllByRole("button", { pressed: true });

  expect(activeSteps).toMatchSnapshot();
});
