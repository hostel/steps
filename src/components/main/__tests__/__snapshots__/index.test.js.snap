// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`check count default steps 1`] = `
<button
  aria-pressed="true"
  class="wrapItem completed"
  data-index="0"
>
  <div
    class="item"
  >
    <p
      class="title"
    >
      Design
    </p>
    <div
      class="circle"
    />
  </div>
</button>
`;

exports[`check count when click last step 1`] = `
Array [
  <button
    aria-pressed="true"
    class="wrapItem completed"
    data-index="0"
  >
    <div
      class="item"
    >
      <p
        class="title"
      >
        Design
      </p>
      <div
        class="circle"
      />
    </div>
  </button>,
]
`;

exports[`check count when click second step 1`] = `
Array [
  <button
    aria-pressed="true"
    class="wrapItem completed nextCompleted"
    data-index="0"
  >
    <div
      class="item"
    >
      <p
        class="title"
      >
        Design
      </p>
      <div
        class="circle"
      />
    </div>
  </button>,
  <button
    aria-pressed="true"
    class="wrapItem completed"
    data-index="1"
  >
    <div
      class="item"
    >
      <p
        class="title"
      >
        Build
      </p>
      <div
        class="circle"
      />
    </div>
  </button>,
]
`;

exports[`should render default active step (first step) 1`] = `
<button
  aria-pressed="true"
  class="wrapItem completed"
  data-index="0"
>
  <div
    class="item"
  >
    <p
      class="title"
    >
      Design
    </p>
    <div
      class="circle"
    />
  </div>
</button>
`;

exports[`should render first and second steps 1`] = `
Array [
  <button
    aria-pressed="true"
    class="wrapItem completed nextCompleted"
    data-index="0"
  >
    <div
      class="item"
    >
      <p
        class="title"
      >
        Design
      </p>
      <div
        class="circle"
      />
    </div>
  </button>,
  <button
    aria-pressed="true"
    class="wrapItem completed"
    data-index="1"
  >
    <div
      class="item"
    >
      <p
        class="title"
      >
        Build
      </p>
      <div
        class="circle"
      />
    </div>
  </button>,
]
`;

exports[`should render first step (b/c clicked last element) 1`] = `
Array [
  <button
    aria-pressed="true"
    class="wrapItem completed"
    data-index="0"
  >
    <div
      class="item"
    >
      <p
        class="title"
      >
        Design
      </p>
      <div
        class="circle"
      />
    </div>
  </button>,
]
`;
